p11-kit/delete-profile.c

```diff
@@ -0,0 +1,224 @@
+		goto cleanup;
+	}
+
+	iter = p11_kit_iter_new (uri, P11_KIT_ITER_WANT_WRITABLE);
+	if (iter == NULL) {
+		p11_message (_("failed to initialize iterator"));
+		goto cleanup;
+	}
+
+	p11_kit_iter_add_filter (iter, &matching, 1);
+	p11_kit_iter_begin (iter, modules);
+	while ((rv = p11_kit_iter_next (iter)) == CKR_OK) {
+		rv = p11_kit_iter_get_attributes (iter, &attr, 1);
+		if (rv != CKR_OK) {
+			p11_message (_("failed to retrieve attribute of an object"));
+			goto cleanup;
+		}
+
+		for (i = 0; i < n_profiles; ++i) {
+			if (profiles[i] == profile_id) {
+				rv = p11_kit_iter_destroy_object (iter);
+				if (rv != CKR_OK)
+					p11_message (_("failed to delete profile: 0x%lX"), profile_id);
+				break;
+			}
+		}
+	}
+
+	ret = 0;
+
+cleanup:
+	p11_kit_iter_free (iter);
+	p11_kit_modules_finalize (modules);
+	p11_kit_modules_release (modules);
+	p11_kit_uri_free (uri);
+
+	return ret;
+}
+
+int
+p11_kit_delete_profile (int argc,
+			char *argv[])
+{
+	int opt;
+	int ret = 2;
+	size_t n_profiles = 0;
+	size_t alloc_profiles = 0;
+	p11_dict *profile_nicks = NULL;
+	CK_PROFILE_ID *tmp = NULL;
+	CK_PROFILE_ID *profiles = NULL;
+	CK_PROFILE_ID profile;
+
+	enum {
+		opt_verbose = 'v',
+		opt_quiet = 'q',
+		opt_help = 'h',
+		opt_profile = 'p',
+	};
+
+	struct option options[] = {
+		{ "verbose", no_argument, NULL, opt_verbose },
+		{ "quiet", no_argument, NULL, opt_quiet },
+		{ "help", no_argument, NULL, opt_help },
+		{ "profile", required_argument, NULL, opt_profile },
+		{ 0 },
+	};
+
+	p11_tool_desc usages[] = {
+		{ 0, "usage: p11-kit delete-profile [--profile profile]... pkcs11:token" },
+		{ opt_profile, "specify the profile to delete" },
+		{ 0 },
+	};
+
+	profile_nicks = p11_constant_reverse (true);
+	if (profile_nicks == NULL) {
+		p11_message (_("failed to allocate memory"));
+		goto cleanup;
+	}
+
+	while ((opt = p11_tool_getopt (argc, argv, options)) != -1) {
+		switch (opt) {
+		case opt_verbose:
+			p11_kit_be_loud ();
+			break;
+		case opt_quiet:
+			p11_kit_be_quiet ();
+			break;
+		case opt_help:
+			p11_tool_usage (usages, options);
+			ret = 0;
+			goto cleanup;
+		case opt_profile:
+			profile = p11_constant_resolve (profile_nicks, optarg);
+			if (profile == CKP_INVALID_ID)
+				profile = strtol (optarg, NULL, 16);
+			if (profile == CKP_INVALID_ID) {
+				p11_message (_("failed to convert profile argument: %s"), optarg);
+				goto cleanup;
+			}
+
+			if (n_profiles >= alloc_profiles) {
+				alloc_profiles += 4;
+				tmp = realloc (profiles, alloc_profiles * sizeof (CK_PROFILE_ID));
+				if (tmp == NULL) {
+					p11_message (_("failed to allocate memory"));
+					goto cleanup;
+				}
+				profiles = tmp;
+			}
+
+			profiles[n_profiles++] = profile;
+			break;
+		case '?':
+			goto cleanup;
+		default:
+			assert_not_reached ();
+			break;
+		}
+	}
+
+	argc -= optind;
+	argv += optind;
+
+	if (argc != 1) {
+		p11_tool_usage (usages, options);
+		goto cleanup;
+	}
+
+	ret = delete_profile (*argv, profiles, n_profiles);
```

Comment:
Should we check at least one profile is specified?

Reply:
probably yes.